lib/services/product-manager.js

```diff
@@ -0,0 +1,102 @@
+import _ from 'lodash'
+import Promise from 'bluebird'
+
+export default class ProductManager {
+  constructor (logger, client) {
+    this.client = client
+    this.logger = logger.child({ service: 'productManager' })
+
+    this.loadBatchCount = 20
+    this.loadConcurrency = 2
+  }
+
+  createProduct (product) {
+    this.logger.debug('Creating product', product)
+
+    return this.client.products
+      .create(product)
+      .then(res => res.body)
+  }
+
+  publishProduct (product) {
+    const actions = [{
+      action: 'publish'
+    }]
+
+    return this.updateProduct(product, actions)
+  }
+
+  updateProduct (product, actions) {
+    const request = {
+      version: product.version,
+      actions
+    }
+
+    return this.client.products
+      .byId(product.id)
+      .update(request)
+  }
+
+  filterOutDuplicateProducts (products) {
+    console.log(products)
+    throw new Error('NOT IMPLEMENTED')
+  }
+
+  async getProductsBySku (skus) {
+    // if skus have more than a certain amount run multiple parallel requests
+    if (skus.length > this.loadBatchCount) {
+      const productBatches = Promise.map(
+        _.chunk(skus, this.loadBatchCount), // load max N skus per one request
+        skuBatch => this.getProductsBySku(skuBatch),
+        { concurrency: this.loadConcurrency } // load products with concurrency
+      )
+
+      return this.filterOutDuplicateProducts(_.flatten(productBatches))
+    }
+
+    const skuPredicate = `${skus.join('","')}`
+    const predicate = `masterVariant(sku IN("${skuPredicate}"))`
+      + ` OR variants(SKU IN("${skuPredicate}")`
+
+    return this.client.productProjections
+      .where(predicate)
+      .fetch()
+      .then(res => res.body.results)
+  }
+
+  getProductById (id) {
+    return this.client.productProjections
+      .staged(true)
+      .where(`id="${id}"`)
+      .fetch()
+      .then(res => res.body.results[0])
+  }
+
+  async deleteProductById (id) {
+    const product = await this.getProductById(id)
+
+    if (product)
+      return this.client.products
+        .byId(product.id)
+        .delete(product.version)
+
+    return Promise.resolve()
+  }
+
+  deleteProduct (product) {
+    return this.client.products
+      .byId(product.id)
+      .delete(product.version)
+  }
+
+  removeVariantsFromProduct (product, variantSkus) {
```

Comment:
How do you think this should be implemented? If we're going to execute the action here, then how are we going to save it in the custom object backup? 😕  The same question applies for method `anonymizeProduct`.

Reply:
you are right .. it will be all in runUpdateActions - see the proposed concept of transactions https://gist.github.com/junajan/b4486d98a3c2b2ddc54a0e8f801d8b3c

But we should also cover promoting of new masterVariant and add a check which will decide whether to run action or not - for example, we remove variant v1 which was a masterVariant so we should promote a new one as a masterVariant.

But before we run this some change can happen and the product won't have any other variant or it won't be a masterVariant anymore so we should not call that promoting of new masterVariant. Or we will get a `concurrent modification` error so we recalculate all actions based on the new state and run it again?

Reply:
I'm thinking should we save the actions or maybe we can just save the product draft? Because in case of an error and you need to refetch from the custom object, you may need to recalculate everything anyway?
Or if we save all the actions, then the benefit is only that we don't need to recalculate and can execute the actions as they are with current version.

Reply:
Ah no, it's not enough in case we delete some variants, we might miss some information.

Reply:
I think that we will have to recalculate everytime we start transaction because the products on API could have been changed so maybe it is a good idea to save just a productDraft and variants which are being removed/moved

Reply:
I wonder why product would change (i am ok with inventories change -> version change)? I think we need to state clearly as part of documentation that any process which changes product data needs to make sure that there are no pending operations in TRANSACTION_CONTAINER and only continue if it is empty. I do not like this dependency but otherwise we might run into weird scenarios:

-> actions [{update P1 and V1 attribute to value 10}]
-> another process would change P1 and V1 attribute to value 20
-> we rerun later actions from 1 step and override to 10 again -> wrong state!

Reply:
Also we should consider that we raise probability of hitting the limit of 16MB per resource if we save to much data.

Reply:
OK, we will state it clearly in the docs what you wrote. However, from the code perspective it seems easier and less code to write if we recalculate the actions from productDraft saved in custom object than saving the actions itself. Because if we save actions, then we need to add different codes to check if this action has happened etc. 
And as to the 16 mb limit, both approaches could reach the limit with the same probability. We discussed with Jan and came to conclusion that we can add compression to the object. https://github.com/commercetools/commercetools-node-variant-reassignment/issues/10

Reply:
We have two options - calculate actions which will be executed in specified order but then if/when the process crashes we will have to rerun those actions again but some of them (eg. removing variant from a product) can be already invalid so we will have to check if the actions is still a valid one. So we have to implement action scheduler for calculating actions and then some action validator.

Or we can take productDraft and removed variants from an unfinished transaction and use scheduler again to calculate valid update actions.

Reply:
I guess it is ok if save product draft.

Reply:
@lojzatran @junajan Please provide issue or clear documentation as stated here: https://github.com/commercetools/commercetools-node-variant-reassignment/pull/9#discussion_r152911526

Reply:
@butenkor I added a documentation issue here https://github.com/commercetools/commercetools-node-variant-reassignment/issues/17